TSPSolver.py

```diff
@@ -2,175 +2,326 @@
+        if nextCity != None:
+          unvisitedCitiesSet.remove(nextCity)
+          route.append(nextCity)
+          currentCity = nextCity
+        else:
+          raise Exception("Unable to visit any city!!")
+      
+      solution = TSPSolution(route)
+      
+      if solution.cost < math.inf:
+        # Found a valid route
+        foundTour = True
+        # Add logic for tracking bssf
+        if bestSolution == None or solution.cost < bestSolution.cost: 
+          count += 1
+          bestSolution = solution
+
+        # Removed old startCity changer
+
+    # Return results
+    end_time = time.time()
+    # Change to bestSolution
+    results['cost'] = bestSolution.cost if foundTour else math.inf
+    results['time'] = end_time - start_time
+    results['count'] = count
+    results['soln'] = bestSolution
+    results['max'], results['total'], results['pruned'] = None, None, None
+    return results
+  
+  
+  
+  ''' <summary>
+    This is the entry point for the branch-and-bound algorithm that you will implement
+    </summary>
+    <returns>results dictionary for GUI that contains three ints: cost of best solution, 
+    time spent to find best solution, total number solutions found during search (does
+    not include the initial BSSF), the best solution found, and three more ints: 
+    max queue size, total number of states created, and number of pruned states.</returns> 
+  '''
+    
+  def branchAndBound( self, time_allowance=60.0 ): # Best Case: O(n^2 * b^n), worst case: O(n!) up to 60 seconds
+    results = {} # O(1)
+    cities = self._scenario.getCities() # O(1)
+    ncities = len(cities) # O(1)
+    ## Get initial BSSF using defaultRandomTour
+    bssf = self.defaultRandomTour(time_allowance)['soln'] # O(n)
+    start_time = time.time() # O(1)
+
+    redCostMatrix = {}
+    indexes = []
+    ## Create an initial reduced cost matrix:  Total = O(n^2)
+    for row in range(ncities): # O(n)
+      indexes.append(row) # O(1)
+      for col in range(ncities): # O(n)
+        redCostMatrix[(row,col)] = cities[row].costTo(cities[col]) # O(1)
+    
+    start = 0
+    route = []
+    route.append(start)
+    ## Could start with any city, I just chose the first one
+    state = TSPState(0, redCostMatrix, indexes.pop(start), indexes, 0, route)
+    heap = []
+    heap.append(state)
+    ## Initialize a heap
+    heapq.heapify(heap) # O(1)
+    ## Pass to a prune and expand function
+    soln, maxSize, totalStates, prunedStates, bssfUpdates = self.findBSSF(heap, bssf, start_time, time_allowance) # n^2 < O() < n! OR 60 seconds
+    
+    end_time = time.time()
+    results['cost'] = soln.cost
+    results['time'] = end_time - start_time
+    results['count'] = bssfUpdates
+    results['soln'] = soln
+    results['max'] = maxSize
+    results['total'] = totalStates
+    results['pruned'] = prunedStates
+    return results
+
+  def findBSSF(self, heap, bssf, start_time, time_allowance): # n^2 * b^n < O() < n! OR 60 seconds
```

Comment:
Can you refactor these extra functions to be inside a TSPBranchAndBound file/class so it isn't taking up a bunch of room here?